### PR TITLE
Plant crossbreeds can now upgrade reagent %s, seeds can now select if they can mutate early.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
No really what do you MEAN YOU CAN GROW PEOPLE?

## About The Pull Request

So, fun fact: plants replace their nutriment with their ground reagent. However, as a result, wheat can only produce a maximum of 4 units of flour, each, per wheat. It makes wheat a massive, MASSIVE pain in the ass to actually produce flour with. So, this PR extends plant cross-pollination to check if an adjacent plant can spread a reagent that it produces more of, so your 4% nutriment wheat can produce 8%, 10%, 15% etc, limited by the higher value of it's adjacent plant's rate.

In addition, seeds now carry a variable early_mutate, enabled by default, but disabled for plants that have truly weird plant products like, say, WHOLE PEOPLE.

## Why It's Good For The Game

This was a feature of the DNA modifier that I hadn't fully considered the ramifications of not replacing, as wheat had to have it's nutriment gene completely removed , necessitating mid-game, R&D locked tools, then selectively breeding endurance of the plant so that it's capable of receiving a new nutriment, THEN and ONLY THEN, can wheat produce more flour. This helps mitigate an exceptionally tedious process within new botany, and even then there are very few plants where you really want to upgrade similar nutrients anyway, but for those situations the time investment shouldn't be 20 minutes.

Additionally, there's a distinct chance that future mutated plants may have more or less sane mutated products, so having this tied to a variable for some of the less sane plant products should prevent crazy situations like, ya know, accidentally getting human byproducts from botany.

## Changelog
:cl:
add: When 2 plants are crossbreeding, if they share a reagent, one plant can now accept and upgrade their production % from the other. 
fix: Cabbage no longer wild-mutates into producing humans as a byproduct
/:cl:

### CODE REVIEW

**Comment by Tlaltecuhtli** [comment]
if you have 2 weath togheter they make more wehat?

**Comment by Galdar02** [comment]
> if you have 2 weath togheter they make more wehat?

Look at ya keyboard before sending, big ol doofus

**Comment by ArcaneMusic** [comment]
> If you have 2 wheat together, do they make more wheat?

This wouldn't impact 2 plants with the same genes if that's what you're asking, think of it more that if a plant would ordinarily fail to add the same reagent to another play, this adjusts the plant's reagent rate instead.

**Comment by ExcessiveUseOfCobblestone** [comment]
he's saying if I have a plant that produces 23% cutegirl and one that produces 10% cutegirl, the 10% one has a chance to convert to 23%.

**Comment by Nebulacrity** [comment]
Sad to see this go, would be nice to have someone remake the bug into a feature. o7

File: code/modules/hydroponics/plant_genes.dm [hunk 1]
```diff
@@ -155,10 +155,17 @@
...
 	for(var/datum/plant_gene/reagent/R in S.genes)
-		if(R.reagent_id == reagent_id)
+		if(R.reagent_id == reagent_id && R.rate <= rate)
 			return FALSE
 	return TRUE
 
+/datum/plant_gene/reagent/proc/try_upgrade_gene(obj/item/seeds/S)
+	for(var/datum/plant_gene/reagent/R in S.genes)
+		if(R.reagent_id == reagent_id && R.rate > rate)
+			rate = R.rate
```
Comment (Rohesie) [inline code/modules/hydroponics/plant_genes.dm:165]: Please, better variable names than `S` and `R`

File: code/modules/hydroponics/seeds.dm [hunk 2]
```diff
@@ -209,17 +211,21 @@
 		product_count = clamp(round(product_count/2),0,5)
 	while(t_amount < product_count)
 		var/obj/item/reagent_containers/food/snacks/grown/t_prod
-		if(instability >= 30 && prob(instability/3) && mutatelist)
+		if(instability >= 30 && prob(instability/3) && mutatelist.len && mutate_early)
```
Comment (Rohesie) [inline code/modules/hydroponics/seeds.dm:214]:
```suggestion
		if(instability >= 30 && prob(instability/3) && length(mutatelist) && mutate_early)
```

File: code/modules/hydroponics/seeds.dm [hunk 3]
```diff
@@ -209,17 +211,21 @@
...
+				continue
+			t_prod = new t_prod(output_loc, src)
+			t_prod.seed = new new_prod
+			t_prod.seed.name = initial(new_prod.name)
+			t_prod.seed.desc = initial(new_prod.desc)
+			t_prod.seed.plantname = initial(new_prod.plantname)
+			t_amount++
+			if(t_prod.seed)
+				//t_prod.seed = new new_prod
+				t_prod.seed.instability = round(instability/2)
```
Comment (Rohesie) [inline code/modules/hydroponics/seeds.dm:227]:
```suggestion
				t_prod.seed.instability = round(instability * 0.5)
```

File: code/modules/hydroponics/plant_genes.dm [hunk 4]
```diff
@@ -155,10 +155,23 @@
...
 	return TRUE
 
+/**
+  * Intends to compare a reagent gene with a set of seeds, and if the seeds contain the same gene, with more production rate, upgrades the rate to the highest of the two.
+  *
+  * Called when plants are crossbreeding, this looks for two matching reagent_ids, where the rates are greater, in order to upgrade.
+  */
+
+/datum/plant_gene/reagent/proc/try_upgrade_gene(obj/item/seeds/seed)
+	for(var/datum/plant_gene/reagent/reagent in seed.genes)
```
Comment (Rohesie) [inline code/modules/hydroponics/plant_genes.dm:169]: Does this list contain any type other than `/datum/plant_gene/reagent` ?

File: code/modules/hydroponics/plant_genes.dm [hunk 5]
```diff
@@ -155,10 +155,23 @@
...
+		if(reagent.reagent_id == reagent_id && reagent.rate > rate)
+			rate = reagent.rate
+			return TRUE
```
Comment (Rohesie) [inline code/modules/hydroponics/plant_genes.dm:172]:
```suggestion
		if(reagent.reagent_id != reagent_id || reagent.rate <= rate)
			continue
		rate = reagent.rate
		return TRUE
```
Prefer early continues.

File: code/modules/hydroponics/seeds.dm [hunk 6]
```diff
@@ -59,6 +59,8 @@
 	var/grafted = FALSE
 	///Type-path of trait to be applied when grafting a plant.
 	var/graft_gene
+	///Determines if the plant should be allowed to mutate early at 30+ instability.
+	var/mutate_early = TRUE
```
Comment (Rohesie) [inline code/modules/hydroponics/seeds.dm:63]: Could you add booleans to a bitflag variable? Something like `seed_flags = MUTATE_EARLY`

File: code/modules/hydroponics/plant_genes.dm (hunk 4 above)
Comment (ArcaneMusic) [reply code/modules/hydroponics/plant_genes.dm:169]: It does, yes. `plant_gene/trait`s as well.

File: code/modules/hydroponics/seeds.dm [hunk 7]
```diff
@@ -59,6 +59,8 @@
 	var/grafted = FALSE
 	///Type-path of trait to be applied when grafting a plant.
 	var/graft_gene
+	///Determines if the plant should be allowed to mutate early at 30+ instability.
+	var/seed_flags = TRUE
```
Comment (Rohesie) [inline code/modules/hydroponics/seeds.dm:63]:
```suggestion
	var/seed_flags = NONE
```

Comment (Rohesie) [inline code/modules/hydroponics/seeds.dm:63]:
```suggestion
	///Bitflags for boolean seed values, such as MUTATE_EARLY
	var/seed_flags = NONE
```
I assume you don't want things to mutate by default.

File: code/__DEFINES/botany.dm [hunk 8]
```diff
@@ -7,3 +7,6 @@
 //Both available scanning modes for the plant analyzer.
 #define PLANT_SCANMODE_STATS		0
 #define PLANT_SCANMODE_CHEMICALS 	1
+
+//Seed flags.
+#define MUTATE_EARLY	(1<<0)
```
Comment (Rohesie) [inline code/__DEFINES/botany.dm:12]:
```suggestion
//Seed flags.
///Determines if the plant should be allowed to mutate early at 30+ instability.
#define MUTATE_EARLY	(1<<0)
```

File: code/modules/hydroponics/seeds.dm [hunk 9]
```diff
@@ -209,17 +211,21 @@
 		product_count = clamp(round(product_count/2),0,5)
 	while(t_amount < product_count)
 		var/obj/item/reagent_containers/food/snacks/grown/t_prod
-		if(instability >= 30 && prob(instability/3) && mutatelist)
+		if(instability >= 30 && prob(instability/3) && length(mutatelist) && (seed_flags & MUTATE_EARLY))
```
Comment (Rohesie) [inline code/modules/hydroponics/seeds.dm:214]:
```suggestion
		if(instability >= 30 && seed_flags & MUTATE_EARLY && length(mutatelist) && prob(instability/3))
```
Try to order checks for cost and relevance. No need to run a probability if the event has no potential to happen, for example.

File: code/modules/hydroponics/seeds.dm (hunk 7 above)
Comment (ArcaneMusic) [reply code/modules/hydroponics/seeds.dm:63]: No, that's the goal here, the issue is that there are specific plants whose product when mutated (Clonepods from cabbage) that will run into issues when mutated, which really made it more akin to an exception. I can go and swap the bigflag over to something like `NO_MUTATION_EARLY` instead in that case

File: code/__DEFINES/botany.dm [hunk 10]
```diff
@@ -7,3 +7,6 @@
 //Both available scanning modes for the plant analyzer.
 #define PLANT_SCANMODE_STATS		0
 #define PLANT_SCANMODE_CHEMICALS 	1
+
+//Seed flags.
+#define NO_MUTATE_EARLY	(1<<0)
```
Comment (Rohesie) [inline code/__DEFINES/botany.dm:12]: Don't do this, avoid negative variables. Not to end up with double-negative stuff like `if(!no_stuff)` which is much more clear as a `if(stuff)`

**Pull request merged**